### PR TITLE
Add Vsc creation

### PR DESCRIPTION
**Pull request by YenguiSeddik**

### CODE REVIEW

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                .operatorActivePowerLimitSide2(8.7F)
+                .droop(1.1F)
+                .angleDroopActivePowerControl(false)
+                .converterStation1(buildConverterStationWithReactiveCapabilityCurve())
+                .converterStation2(buildConverterStationWithMinMaxReactiveLimits())
+                .build();
+    }
+
+    private ConverterStationCreationInfos buildConverterStationWithMinMaxReactiveLimits() {
+        var test = getNetwork().getVoltageLevelStream().toList();
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:58]: remove unused test variable

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                .build();
+    }
+
+    @Override
+    protected void assertAfterNetworkModificationCreation() {
+        var list = getNetwork().getHvdcLineStream().toList();
+        assertNotNull(getNetwork().getHvdcLine("vsc1"));
+
+        assertEquals(1, getNetwork().getVoltageLevel("v1").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId1")).count());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:131]: change 'transformer' variable name to converterStation

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+        assertEquals(1, getNetwork().getVoltageLevel("v2").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId2")).count());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:134]: change 'transformer' variable name to converterStation

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Override
+    protected void assertAfterNetworkModificationDeletion() {
+        assertNull(getNetwork().getHvdcLine("vsc1"));
+
+        assertEquals(0, getNetwork().getVoltageLevel("v1").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId1")).count());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:142]: change 'transformer' variable name to converterStation

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        assertEquals(0, getNetwork().getVoltageLevel("v2").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId2")).count());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:145]: change 'transformer' variable name to converterStation

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    protected void assertAfterNetworkModificationCreation() {
+        var list = getNetwork().getHvdcLineStream().toList();
+        assertNotNull(getNetwork().getHvdcLine("vsc1"));
+
+        assertEquals(1, getNetwork().getVoltageLevel("v1").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId1")).count());
+
+        assertEquals(1, getNetwork().getVoltageLevel("v2").getVscConverterStationStream()
+                .filter(transformer -> transformer.getId().equals("stationId2")).count());
+    }
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java:135]: Add more checks on the the vsc characteristics, and on the vsc converter station 1 and station 2 characteristics ...

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                CREATE_VSC_ERROR,
+                modificationInfos.getEquipmentId(),
+                "Vsc");
+    }
+
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        VscConverterStation converterStation1 = modificationInfos.getConverterStation1() == null ?
+                null :
+                createConvertStation(network, modificationInfos.getConverterStation1(), subReporter, "Converter station 1");
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java:78]: createConverterStation

File: src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+                                                     String logFieldName) {
+        Reporter converterStationReporter = subReporter.createSubReporter("converterStationCreated", logFieldName);
+        List<Report> converterStationReports = new ArrayList<>();
+        VoltageLevel voltageLevel = ModificationUtils.getInstance().getVoltageLevel(network, converterStationCreationInfos.getVoltageLevelId());
+        VscConverterStation vscConverterStation = voltageLevel.getTopologyKind() == TopologyKind.NODE_BREAKER ?
+                createConvertStationInNodeBreaker(network, voltageLevel, converterStationCreationInfos, converterStationReports, converterStationReporter) :
+                createConvertStationInBusBreaker(voltageLevel, converterStationCreationInfos, converterStationReports, converterStationReporter);
+
+        converterStationReporter.report(Report.builder()
+                .withKey("converterStationCreated")
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java:182]: you cannot use the same key here for converter station 1 and converter station 2

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/dto/ConverterStationCreationInfos.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ * @author Seddik Yengui <seddik.yengui at rte-france.com>
+ */
+
+@SuperBuilder
+@NoArgsConstructor
+@Getter
+@Setter
+@ToString(callSuper = true)
+@Schema(description = "Converter station creation")
+public class ConverterStationCreationInfos implements ReactiveLimitsHolderInfos {
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/dto/ConverterStationCreationInfos.java]: done

File: src/main/java/org/gridsuite/modification/server/dto/VscCreationInfos.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    private Double dcNominalVoltage;
+
+    @Schema(description = "DC resistance")
+    private Double dcResistance;
+
+    @Schema(description = "Maximum active power ")
+    private Double maximumActivePower;
+
+    @Schema(description = "Operator active power limit (Side1->Side2)")
+    private Float operatorActivePowerLimitSide1;
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/dto/VscCreationInfos.java]: done

File: src/main/java/org/gridsuite/modification/server/dto/VscCreationInfos.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Schema(description = "Operator active power limit (Side2->Side1)")
+    private Float operatorActivePowerLimitSide2;
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/dto/VscCreationInfos.java]: done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ConverterStationCreationEntity.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+/**
+ * @author Seddik Yengui <seddik.yengui at rte-france.com>
+ */
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Entity
+@Table(name = "converterStationCreationEntity")
+public class ConverterStationCreationEntity {
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ConverterStationCreationEntity.java]: done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ConverterStationCreationEntity.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+/**
+ * @author Seddik Yengui <seddik.yengui at rte-france.com>
+ */
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Entity
+@Table(name = "converterStationCreationEntity")
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ConverterStationCreationEntity.java]: done

File: src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java (hunk 1 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/modification/server/modifications/VscCreationTest.java]: done

File: src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+        }
+
+        if (modificationInfos.getDroop() != null) {
+            angleDroopActivePowerControlReports.add(ModificationUtils.getInstance().buildCreationReport(modificationInfos.getDroop(), "Droop"));
+        }
+
+        ModificationUtils.getInstance().reportModifications(setPointsReporter, angleDroopActivePowerControlReports, "vscAngleDroop", "Angle droop active power control");
+    }
+
+    private VscConverterStation createConvertStation(Network network,
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/VscCreation.java:165]: createConverterStation

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Review by Walid-Sahnoun — APPROVED** [review]
code ok
test ok

**Pull request merged**